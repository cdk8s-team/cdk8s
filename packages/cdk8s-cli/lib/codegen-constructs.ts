import { JSONSchema4 } from "json-schema";
import { CodeMaker } from 'codemaker';
import { TypeGenerator } from "./codegen-types";

export interface GroupVersionKind {
  readonly group: string;
  readonly kind: string;
  readonly version: string
}

export const X_GROUP_VERSION_KIND = 'x-kubernetes-group-version-kind';

/**
 * Generates a construct for an API object defined by `def`.
 */
export function emitConstructForApiObject(code: CodeMaker, schema: JSONSchema4, def: JSONSchema4) {
  const objectNames = def[X_GROUP_VERSION_KIND] as GroupVersionKind[];
  if (!objectNames) {
    throw new Error(`object must include a ${X_GROUP_VERSION_KIND} key`);
  }

  const objectName = objectNames[0];
  if (!objectName) {
    throw new Error(`no object name`);
  }

  const groupPrefix = objectName.group ? `${objectName.group.toLocaleLowerCase().replace(/\./g, '-')}-` : '';
  const baseName = `${groupPrefix}${objectName.kind.toLocaleLowerCase()}-${objectName.version.toLocaleLowerCase()}`;

  if (!def.properties?.metadata) {
    console.error(`warning: no "metadata", skipping ${baseName}`);
    return;
  }

  const sourceFile = `${baseName}.ts`;
  const optionsStructName = `${objectName.kind}Options`;

  const typeGenerator = new TypeGenerator(schema);

  emitFile();

  function emitFile() {
    code.openFile(sourceFile);
    code.line(`// generated by cdk8s`);
    code.line();

    code.line(`import { ApiObject } from 'cdk8s';`);
    code.line(`import { Construct } from '@aws-cdk/core';`);
    code.line();
  
    emitOptionsStruct();

    code.line();

    emitConstruct();
    code.line();

    typeGenerator.generate(code);
  
    code.closeFile(sourceFile);
  }

  function emitOptionsStruct() {
    const copy: JSONSchema4 = { ...def };
    copy.properties = copy.properties || {};
    delete copy.properties!.apiVersion;
    delete copy.properties!.kind;
    delete copy.properties!.status;

    typeGenerator.addType(optionsStructName, copy);
  }
  
  function emitConstruct() {
    code.line('/**');
    code.line(` * ${ def?.description }`);
    code.line(` */`);
    code.openBlock(`export class ${objectName.kind} extends ApiObject`);

    emitInitializer();
  
    code.closeBlock();
  }

  function emitInitializer() {
    code.openBlock(`public constructor(scope: Construct, ns: string, options: ${optionsStructName})`);
    emitInitializerSuper();

    code.closeBlock();
  }

  function emitInitializerSuper() {
    const groupPrefix = objectName.group ? `${objectName.group}/` : '';
    code.open(`super(scope, ns, {`);
    code.line(`...options,`);
    code.line(`kind: '${objectName.kind}',`);
    code.line(`apiVersion: '${groupPrefix}${objectName.version}',`);
    code.close(`});`);    
  }
}

/**
 * Returns all schema definitions for API objects (objects that have the 'x-kubernetes-group-version-kind' annotation)
 */
export function findApiObjectDefinitions(schema: JSONSchema4) {
  const result = new Array<JSONSchema4>();
  for (const def of Object.values(schema.definitions || { })) {
    const kinds = def[X_GROUP_VERSION_KIND];
    if (!kinds) {
      continue;
    }

    result.push(def);
  }

  return result;
}