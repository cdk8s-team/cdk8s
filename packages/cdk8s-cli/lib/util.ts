import * as http from 'http';
import * as https from 'https';
import { spawn, SpawnOptions } from 'child_process';
import * as fs from 'fs-extra';
import * as os from 'os';
import * as path from 'path';

export async function shell(program: string, args: string[] = [], options: SpawnOptions = { }) {
  return new Promise((ok, ko) => {
    const child = spawn(program, args, { stdio: 'inherit', ...options });
    child.once('error', ko);
    child.once('exit', code => {
      if (code === 0) { return ok(); }
      else { return ko(new Error(`non-zero exit code ${code}`)); }
    });
  });
}

export async function withTempDir(dirname: string, closure: () => Promise<void>) {
  const prevdir = process.cwd();
  const parent = await fs.mkdtemp(path.join(os.tmpdir(), 'cdk8s.'));
  const workdir = path.join(parent, dirname);
  await fs.mkdirp(workdir);
  try {
    process.chdir(workdir);
    await closure();
  } finally {
    process.chdir(prevdir);
    await fs.remove(parent);
  }
}

async function get(url: string, protocol: typeof http | typeof https = https): Promise<string> {
  return new Promise((ok, ko) => {
    const req = protocol.get(url, async res => {
      switch(res.statusCode) {
        case 200: {
          const data = new Array<Buffer>();
          res.on('data', chunk => data.push(chunk));
          res.once('end', () => ok(Buffer.concat(data).toString('utf-8')));
          res.once('error', ko);
          break;
        }

        case 301:
        case 302: {
          if (res.headers.location) {
            await ok(get(res.headers.location, protocol));
          }
          break;
        }

        default: {
          throw new Error(`${res.statusMessage}: ${url}`);
        }
      }
    });

    req.once('error', ko);
    req.end();
  });
}

export async function httpGet(url: string): Promise<string> {
  return get(url, http)
}

export async function httpsGet(url: string): Promise<string> {
  return get(url)
}
