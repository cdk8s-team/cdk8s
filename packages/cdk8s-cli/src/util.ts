import * as http from 'http';
import * as https from 'https';
import { spawn, SpawnOptions } from 'child_process';
import { parse } from 'url';
import * as fs from 'fs-extra';
import * as os from 'os';
import * as path from 'path';

export async function shell(program: string, args: string[] = [], options: SpawnOptions = { }) {
  const command = `"${program} ${args.join(' ')}" at ${path.resolve(options.cwd ?? '.')}`;
  return new Promise((ok, ko) => {
    const child = spawn(program, args, { stdio: 'inherit', ...options });
    child.once('error', err => {
      throw new Error(`command ${command} failed: ${err}`);
    });
    child.once('exit', code => {
      if (code === 0) {
        return ok();
      }
      else {
        return ko(new Error(`command ${command} returned a non-zero exit code ${code}`));
      }
    });
  });
}

export async function mkdtemp(closure: (dir: string) => Promise<void>) {
  const workdir = await fs.mkdtemp(path.join(os.tmpdir(), 'cdk8s-'));
  try {
    await closure(workdir);
  } finally {
    await fs.remove(workdir);
  }
}

export async function download(url: string): Promise<string> {
  let client: typeof http | typeof https;
  const proto = parse(url).protocol;

  if (!proto || proto === 'file:') {
    return await fs.readFile(url, 'utf-8');
  }

  switch (proto) {
    case 'https:':
      client = https;
      break;

    case 'http:':
      client = http;
      break;

    default:
      throw new Error(`unsupported protocol ${proto}`);
  }

  return new Promise((ok, ko) => {
    const req = client.get(url, res => {
      switch(res.statusCode) {
        case 200: {
          const data = new Array<Buffer>();
          res.on('data', chunk => data.push(chunk));
          res.once('end', () => ok(Buffer.concat(data).toString('utf-8')));
          res.once('error', ko);
          break;
        }

        case 301:
        case 302: {
          if (res.headers.location) {
            ok(download(res.headers.location));
          }
          break;
        }

        default: {
          throw new Error(`${res.statusMessage}: ${url}`);
        }
      }
    });

    req.once('error', ko);
    req.end();
  });
}
