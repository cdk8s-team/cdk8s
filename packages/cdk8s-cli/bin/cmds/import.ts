import * as yargs from 'yargs';
import { generateAllApiObjects, DEFAULT_API_VERSION } from '../../lib/import';

const K8S_SPEC = 'k8s';
const DEFAULT_OUTDIR = 'imports';

class Command implements yargs.CommandModule {
  public readonly command = 'import SPEC';
  public readonly describe = 'Imports API objects to your app by generating constructs';
  public readonly aliases = [ 'gen', 'import', 'generate' ];

  public readonly builder = (args: yargs.Argv) => args
    .positional('SPEC', {  desc: `The API spec to import (only "k8s" is currently supported)` })
    .demand('SPEC')
    .example(`cdk8s import k8s`, `imports Kubenretes API objects to imports/k8s.ts`)
    .example(`cdk8s import k8s@1.13.0`, `imports a specific version of the Kubenretes API`)

    .option('output', { type: 'string', desc: 'output directory', default: DEFAULT_OUTDIR, alias: 'o' })
    .option('include', { type: 'array', desc: 'types to select instead of the default which is latest stable version' })
    .option('exclude', { type: 'array', desc: 'do not import types that match these regular expressions. They will be represented as the "any" type' });

  public async handler(argv: any) {
    const spec = argv.spec;
    const [ src, specver ] = spec.split('@');

    const apiVersion = specver ?? DEFAULT_API_VERSION;

    if (src !== K8S_SPEC) {
      throw new Error(`currently we only support the source ${K8S_SPEC}`);
    }

    await generateAllApiObjects(argv.output, {
      apiVersion,
      include: argv.include,
      exclude: argv.exclude
    });

    console.error(`k8s@${apiVersion} imported to ${argv.output}/k8s.ts`);
  }
}


module.exports = new Command();
