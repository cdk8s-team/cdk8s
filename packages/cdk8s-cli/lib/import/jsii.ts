import * as fs from 'fs-extra';
import * as path from 'path';
import { shell } from '../util';

export interface JsiiCompileOptions {
  /**
   * Name of the module.
   */
  readonly name: string;

  /**
   * Optional prefix for the module for custom-naming.
   */
  readonly moduleNamePrefix?: string;

  /**
   * Name of the main module file without suffix.
   */
  readonly main: string;

  /**
   * Whether to print stdout.
   * @default false
   */
  readonly stdout?: boolean;
}

/**
 * Compiles the source files in `workdir` with jsii.
 */
export async function jsiiCompile(workdir: string, options: JsiiCompileOptions) {
  const stdout = options.stdout ?? false;
  const name = (options.name ?? 'dummy').replace(/[^a-z0-9]/gi, '_').toLocaleLowerCase(); // slugify
  const compiler = require.resolve('jsii/bin/jsii');
  const args = [ '--silence-warnings', 'reserved-word' ];

  const main = options.main;
  const moduleNamePrefix = options.moduleNamePrefix;

  const modules = [ 
    'constructs', 
    'cdk8s',
    '@types/node',
  ];

  const pkg = {
    name,
    version: '0.0.0',
    author: "dummy@dummy.com",
    main: moduleNamePrefix ? `${moduleNamePrefix}-${main}.js` : `${main}.js`,
    types: moduleNamePrefix ? `${moduleNamePrefix}-${main}.d.ts` : `${main}.d.ts`,
    license: 'Apache-2.0',
    repository: { url: 'http://repo', type: 'git' },
    jsii: {
      outdir: "dist",
      targets: {
        python: {
          distName: moduleNamePrefix ? `${moduleNamePrefix}.${main}` : main,
          module: moduleNamePrefix ? `${moduleNamePrefix}.${main}` : main
        }
      }
    },    
    dependencies: {
      "constructs": "*",
      "cdk8s": "*",
    },
    peerDependencies: {
      "constructs": "*",
      "cdk8s": "*",
    }
  };

  for (const mod of modules) {
    const sourcedir = path.dirname(require.resolve(`${mod}/package.json`));
    await fs.mkdirp(path.join(workdir, path.join('node_modules', path.dirname(mod))));
    await fs.ensureSymlink(sourcedir, path.join(workdir, 'node_modules', mod));
  }

  await fs.writeFile(path.join(workdir, 'package.json'), JSON.stringify(pkg, undefined, 2));

  await shell(compiler, args, { 
    cwd: workdir,
    stdio: [ 'inherit', stdout ? 'inherit' : 'ignore', 'inherit' ]
  });
}