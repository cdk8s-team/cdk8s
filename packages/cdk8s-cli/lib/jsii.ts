import * as fs from 'fs-extra';
import * as path from 'path';
import { shell } from './util';

/**
 * Compiles the source files in `workdir` with jsii.
 */
export async function jsiiCompile(workdir: string) {
  const compiler = require.resolve('jsii/bin/jsii');
  const args = [ '--silence-warnings', 'reserved-word' ];

  const modules = [ 
    '@aws-cdk/core', 
    '@aws-cdk/cx-api', 
    'cdk8s',
    '@types/node'
  ];

  const pkg = {
    name: 'dummy',
    version: '0.0.0',
    author: "dummy@dummy.com",
    main: "k8s.js",
    types: "k8s.d.ts",
    license: 'Apache-2.0',
    repository: { url: 'http://repo', type: 'git' },
    jsii: {
      outdir: "dist",
      targets: {
        python: {
          distName: "k8s",
          module: "k8s"
        }
      }
    },    
    dependencies: {
      "@aws-cdk/core": "*",
      "cdk8s": "*",
    },
    peerDependencies: {
      "@aws-cdk/core": "*",
      "cdk8s": "*",
    }
  };

  for (const mod of modules) {
    const sourcedir = path.dirname(require.resolve(`${mod}/package.json`));
    await fs.mkdirp(path.join(workdir, path.join('node_modules', path.dirname(mod))));
    await fs.symlink(sourcedir, path.join(workdir, 'node_modules', mod));
  }

  await fs.writeFile(path.join(workdir, 'package.json'), JSON.stringify(pkg, undefined, 2));

  await shell(compiler, args, { 
    cwd: workdir 
  });
}