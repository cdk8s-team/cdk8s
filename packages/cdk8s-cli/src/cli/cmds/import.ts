import * as yargs from 'yargs';
import { readConfigSync, ImportSpec } from '../../config';
import { importDispatch } from '../../import/dispatch';

const config = readConfigSync();

const DEFAULT_OUTDIR = 'imports';
const LANGUAGES = [ 'typescript', 'python', 'java' ];

class Command implements yargs.CommandModule {
  public readonly command = 'import [SPEC]';
  public readonly describe = 'Imports API objects to your app by generating constructs.';
  public readonly aliases = [ 'gen', 'import', 'generate' ];

  public readonly builder = (args: yargs.Argv) => args
    .positional('SPEC', { default: config.imports, desc: 'import spec with the syntax [NAME:=]SPEC where NAME is an optional module name and supported SPEC are: k8s, crd.yaml, https://domain/crd.yaml).', array: true })
    .example('cdk8s import k8s', 'Imports Kubernetes API objects to imports/k8s.ts')
    .example('cdk8s import k8s@1.13.0', 'Imports a specific version of the Kubernetes API')
    .example('cdk8s import jenkins.io_jenkins_crd.yaml', 'Imports constructs for the Jenkins custom resource definition from a file')
    .example('cdk8s import mattermost:=mattermost_crd.yaml', 'Imports constructs for the mattermost cluster custom resource definition using a custom module name')

    .option('output', { default: DEFAULT_OUTDIR, type: 'string', desc: 'Output directory', alias: 'o' })
    .option('include', { type: 'array', desc: 'Types to select instead of the default which is latest stable version (only for "k8s")' })
    .option('exclude', { type: 'array', desc: 'Do not import types that match these regular expressions. They will be represented as the "any" type (only for "k8s")' })
    .option('language', { default: config.language, demand: true, type: 'string', desc: 'Output programming language', alias: 'l', choices: LANGUAGES });

  public async handler(argv: any) {
    let specs = Array.isArray(argv.spec) ? argv.spec : [ argv.spec ];

    specs = specs.filter((spec: string) => spec != null).map((spec: string) => parseImports(spec));

    await importDispatch(specs, argv, {
      outdir: argv.output,
      targetLanguage: argv.language,
    });
  }
}

function parseImports(spec: string): ImportSpec {
  const splitImport = spec.split(':=');

  // crd.yaml
  // url.com/crd.yaml
  if (splitImport.length === 1) {
    return {
      source: spec,
    }
  }

  // crd=crd.yaml
  // crd=url.com/crd.yaml
  if (splitImport.length === 2) {
    return {
      moduleNamePrefix: splitImport[0],
      source: splitImport[1],
    }
  }

  throw new Error('Unable to parse import specification. Syntax is [NAME:=]SPEC');
}

module.exports = new Command();
