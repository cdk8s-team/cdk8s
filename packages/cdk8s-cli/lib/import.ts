import * as https from 'https';
import * as fs from 'fs-extra';
import * as path from 'path';
import { promises } from 'fs';
import { emitConstructForApiObject, findApiObjectDefinitions, selectApiObjects } from '../lib/codegen-constructs';
import { CodeMaker } from 'codemaker';
import { JSONSchema4 } from 'json-schema';
import { TypeGenerator } from './codegen-types';
import { withTempDir, shell } from './util';
import { jsiiCompile } from './jsii';

export const DEFAULT_API_VERSION = '1.14.0';

export enum Language {
  TYPESCRIPT = 'typescript',
  PYTHON = 'python',
  DOTNET = 'dotnet',
  JAVA = 'java',
}

export const LANGUAGES = [ Language.TYPESCRIPT, Language.PYTHON ];

export interface Options {
  /**
   * Output programming language.
   */
  readonly language: Language;

  /**
   * The API version to generate.
   */
  readonly apiVersion?: string;

  /**
   * FQNs of API object types to select instead of selecting the latest stable
   * version.
   * 
   * @default - selects the latest stable version from each API object
   */
  readonly include?: string[];

  /**
   * Do not import these types. Instead, represent them as "any".
   * 
   * @default - include all types that derive from the root types.
   */
  readonly exclude?: string[];
}

export async function generateAllApiObjects(outdir: string, options: Options) {
  outdir = path.resolve(outdir);

  if (options.language === Language.TYPESCRIPT) {
    await generateApiObjectsTypeScript(outdir, options);
    console.log(`${outdir}/k8s.ts`);
    return;
  }

  // this is not typescript, so we generate in a staging directory and harvest the code
  await withTempDir('k8s', async () => {
    await generateApiObjectsTypeScript('.', options);
    await jsiiCompile('.');

    const pacmak = require.resolve('jsii-pacmak/bin/jsii-pacmak');
    await shell(pacmak, [ '--target', options.language, '--code-only' ]);
    await harvestCode(options.language, outdir);
  });
}

async function harvestCode(language: Language, outdir: string) {
  switch (language) {
    case Language.TYPESCRIPT:
      throw new Error('no op for typescript');

    case Language.PYTHON:
      await harvestPython();
      break;

    default:
      throw new Error(`unsupported language ${language} (yet)`);
  }

  async function harvestPython() {
    const target = path.join(outdir, 'k8s');
    await fs.move('dist/python/src/k8s', target, { overwrite: true });
    console.error(target);
  }
}

async function generateApiObjectsTypeScript(outdir: string, options: Options) {
  const code = new CodeMaker();
  code.indentation = 2;

  const schema = await downloadSchema(options.apiVersion ?? DEFAULT_API_VERSION);
  const map = findApiObjectDefinitions(schema);

  const topLevelObjects = selectApiObjects(map, { include: options.include });

  code.openFile('k8s.ts');
  code.line(`// generated by cdk8s`);
  code.line(`import { ApiObject } from 'cdk8s';`);
  code.line(`import { Construct } from '@aws-cdk/core';`);
  code.line();

  const typeGenerator = new TypeGenerator(schema, { exclude: options.exclude });

  for (const o of topLevelObjects) {
    emitConstructForApiObject(code, typeGenerator, o);
  }

  typeGenerator.generate(code);

  code.closeFile('k8s.ts');

  await promises.mkdir(outdir, { recursive: true });
  await code.save(outdir);  
}

async function downloadSchema(apiVersion: string) {
  const output = await httpsGet(`https://kubernetesjsonschema.dev/v${apiVersion}/_definitions.json`);
  return JSON.parse(output) as JSONSchema4;
}

async function httpsGet(url: string): Promise<string> {
  return new Promise((ok, ko) => {
    const req = https.get(url, res => {
      if (res.statusCode !== 200) {
        throw new Error(`${res.statusMessage}: ${url}`);
      }
      const data = new Array<Buffer>();
      res.on('data', chunk => data.push(chunk));
      res.once('end', () => ok(Buffer.concat(data).toString('utf-8')));
      res.once('error', ko);
    });

    req.once('error', ko);
    req.end();
  });
}
