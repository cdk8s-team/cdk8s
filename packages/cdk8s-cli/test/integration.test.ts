import { importKubernetesApi, Language } from "../lib/import";
import { jsiiCompile } from "../lib/jsii";
import { promises as fs } from 'fs';
import { withTempDir } from "../lib/util";

jest.setTimeout(60_000);

test('generate and compile 1.14.0', async () => {
  await withTempDir('import-k8s', async () => {
    const workdir = '.';
    await importKubernetesApi(workdir, {
      language: Language.TYPESCRIPT,
      apiVersion: '1.14.0' 
    });
    
    await jsiiCompile(workdir, { stdout: true});
  
    const manifest = JSON.parse(await fs.readFile('.jsii', 'utf-8'));
    expect(manifest).toMatchSnapshot();
  });
});