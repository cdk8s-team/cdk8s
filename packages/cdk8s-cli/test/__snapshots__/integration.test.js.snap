// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate and compile 1.14.0 1`] = `
Object {
  "author": Object {
    "name": "dummy@dummy.com",
    "roles": Array [
      "author",
    ],
  },
  "dependencies": Object {
    "@aws-cdk/core": "*",
    "cdk8s": "*",
  },
  "dependencyClosure": Object {
    "@aws-cdk/core": Object {
      "targets": Object {
        "dotnet": Object {
          "assemblyOriginatorKeyFile": "../../key.snk",
          "iconUrl": "https://raw.githubusercontent.com/aws/aws-cdk/master/logo/default-256-dark.png",
          "namespace": "Amazon.CDK",
          "packageId": "Amazon.CDK",
          "signAssembly": true,
        },
        "java": Object {
          "maven": Object {
            "artifactId": "core",
            "groupId": "software.amazon.awscdk",
          },
          "package": "software.amazon.awscdk.core",
        },
        "js": Object {
          "npm": "@aws-cdk/core",
        },
        "python": Object {
          "distName": "aws-cdk.core",
          "module": "aws_cdk.core",
        },
      },
    },
    "@aws-cdk/cx-api": Object {
      "targets": Object {
        "dotnet": Object {
          "assemblyOriginatorKeyFile": "../../key.snk",
          "iconUrl": "https://raw.githubusercontent.com/aws/aws-cdk/master/logo/default-256-dark.png",
          "namespace": "Amazon.CDK.CXAPI",
          "packageId": "Amazon.CDK.CXAPI",
          "signAssembly": true,
        },
        "java": Object {
          "maven": Object {
            "artifactId": "cdk-cx-api",
            "groupId": "software.amazon.awscdk",
          },
          "package": "software.amazon.awscdk.cxapi",
        },
        "js": Object {
          "npm": "@aws-cdk/cx-api",
        },
        "python": Object {
          "distName": "aws-cdk.cx-api",
          "module": "aws_cdk.cx_api",
        },
      },
    },
    "cdk8s": Object {
      "targets": Object {
        "dotnet": Object {
          "namespace": "Org.Cdk8s",
          "packageId": "Org.Cdk8s",
        },
        "java": Object {
          "maven": Object {
            "artifactId": "cdk8s",
            "groupId": "org.cdk8s",
          },
          "package": "org.cdk8s",
        },
        "js": Object {
          "npm": "cdk8s",
        },
        "python": Object {
          "distName": "cdk8s",
          "module": "cdk8s",
        },
      },
    },
  },
  "description": "dummy",
  "fingerprint": "HD4O8JPi7/yzcp+ASh//v9CjPwF8WUUjrfUevSbCO9A=",
  "homepage": "http://repo",
  "jsiiVersion": "1.0.0 (build eb2db9d)",
  "license": "Apache-2.0",
  "name": "dummy",
  "repository": Object {
    "type": "git",
    "url": "http://repo",
  },
  "schema": "jsii/0.10.0",
  "targets": Object {
    "js": Object {
      "npm": "dummy",
    },
    "python": Object {
      "distName": "index",
      "module": "index",
    },
  },
  "types": Object {},
  "version": "0.0.0",
}
`;
